### ISSUE TITLE
残差块中只用一个卷积么？

### ISSUE THREAD

**Issue by changshuangzhao** [zhixuanli/StereoNet#1]
你好， 最近在看立体匹配，很高兴能使用你的代码，有一个问题，残差块中只用一个卷积么？

**Comment by zhixuanli** [comment]
Sorry I'm a beginner on ResNet, and according to [https://pytorch.org/docs/stable/_modules/torchvision/models/resnet.html#resnet101](https://pytorch.org/docs/stable/_modules/torchvision/models/resnet.html#resnet101)
, there indeed should exists two conv2d.

Thank you so much!

**Comment by zhixuanli** [comment]
Yes, you are right about this. I wrote the soft argmin referencing this :
[https://github.com/liuruijin17/GCNet-tensorflow/blob/35127e7722edca369835bb433c2f17f41111e959/critic_network.py#L8](https://github.com/liuruijin17/GCNet-tensorflow/blob/35127e7722edca369835bb433c2f17f41111e959/critic_network.py#L8)

But that code doesn't consider the batchsize, so "dim=0" is used.

I have fixed this in my code.

-----------------------------------------------------------------------

And the experiment result now is :
```
Average ACC = 0.683487
Average EPE = 6.931608
```
1 pixel error accurancy here.

Though this model is not pretrained on SceneFlow dataset, but I don't believe that only training and testing on KITTI 2015 will just achieve this bad performance.

I'll check again each detail of my code, and then try to pretrain on the SceneFlow dataset.

And thanks again! You are really thoughtful and enthusiastic!

**Comment by changshuangzhao** [comment]
我自己也实现了一个，cost volume直接用的PSPNet中的形式，学校服务器比较紧张，还没有来得急训练。https://github.com/changshuangzhao/StereNet，有问题一起交流！！

**Comment by zhixuanli** [comment]
I'm trying to modify the [PSMNet](https://github.com/JiaRenChang/PSMNet) now.
Hope there can be some good results.

I will contact you if there's any progress!

**Comment by changshuangzhao** [comment]
你好，进展的怎么样？我利用自己编的代码，在SceneFlow数据集上(包括flyingthing3D， Monkaa, Driving)训练，已经训练了三天，配置如下：
RMSprop（初始学习率0.001，指数衰减0.9），
权重衰减： 0.0001
图片大小是960x540, 3次下采样，
没有进行裁剪，利用了归一化，
训练loss：smooth_l1_loss
测试loss：EPE
批次：1
三天才训练10轮（太慢了）得到的结果如下：
Number of model parameters: 398816
This is 1-th epoch
total training loss = 10.123
total testing loss epe = 7.169
This is 2-th epoch
total training loss = 7.712
total testing loss epe = 6.757
This is 3-th epoch
total training loss = 6.997
total testing loss epe = 6.717
This is 4-th epoch
total training loss = 6.572
total testing loss epe = 6.159
This is 5-th epoch
total training loss = 6.260
total testing loss epe = 6.436
This is 6-th epoch
total training loss = 6.053
total testing loss epe = 5.855
This is 7-th epoch
total training loss = 5.859
total testing loss epe = 6.019
This is 8-th epoch
total training loss = 5.705
total testing loss epe = 5.523
This is 9-th epoch
total training loss = 5.544
total testing loss epe = 5.337
This is 10-th epoch
total training loss = 5.435
total testing loss epe = 5.277
This is 11-th epoch
total training loss = 5.314
total testing loss epe = 5.542
虽然测试和训练都在下降，达到论文结果，可能性不太大，这个还有必要在跑下去了么？
我可以看看你跑的训练曲线么？

**Comment by zhixuanli** [comment]
你好，昨天我也做了在SceneFlow数据集上的预训练，但是昨天的时候，Flying things数据集那部分还没下载下来，刚才才终于下载完毕。

所以先说一下昨天的情况，在另外两部分Monkaa, Driving上进行了预训练和测试，
```
optimizer = optim.Adam(model.parameters(), lr=0.001, betas=(0.9, 0.999))
```
学习率恒定为0.001,10个epoch，图片读取方式完全使用了listflowfile.py中的方式，可以看我现在代码仓库中的sceneflow_pretrain.py，只改动了很小的一部分。
最后loss记得是有在一直下降，但是没有你的loss那么小。
现在正在尝试在完整的SceneFlow数据集上训练。有结果了就马上发给你。

除此之外有一个问题想和你讨论：

PSMNet代码的作者在finetune的文件里，train部分，定义的loss为
```
mask = (disp_true > 0)
mask.detach_()

loss = F.smooth_l1_loss(output[mask], disp_true[mask], size_average=True)
```

这里为什么要用mask。实际上作者在finetune的test过程中也使用了mask，但是一直不是很理解为什么只对比在ground truth中有数值的部分，而对于是0的数值则不去计算它。难道GT中数值为0的部分所对应的网络的预测图中相应位置的元素，是不重要的吗。

此外，PSM代码中，mask在预训练和finetune过程中的定义是不同的

------------------------
我会尽快在完整数据集上预训练好然后告诉你结果。

**Comment by changshuangzhao** [comment]
我是这样理解的，KITTI数据是稀疏的，导致一部分图像没有真实的视差，设定GT=0。
GT为0的像素代表没有对应的视差值，在测试阶段不需要对其进行评测

**Comment by zhixuanli** [comment]
有道理，这下就说的通了:)

**Comment by zhixuanli** [comment]
我在sceneflow上预训练完并在Kitti15上进行了finetune，但是小于1 pixel error acc为90.54，没有论文里高。

我改了优化器为Adam，并设置前200轮为0.001，后100轮为0.0001。那种指数衰减的方式不知道为什么不是很好用，可能是因为衰减的太快了。但是找到合适的衰减速率也有些麻烦。

我再继续尝试改进，保持交流

**Comment by changshuangzhao** [comment]
scene flow训练了20轮，我就停了下来，因为初始学习率0.001，衰减率0.9，20轮后学习率太小了，
我没有在kitti上测试，
umber of model parameters: 398816
This is 1-th epoch
total training loss = 10.123
total testing loss epe = 7.169
This is 2-th epoch
total training loss = 7.712
total testing loss epe = 6.757
This is 3-th epoch
total training loss = 6.997
total testing loss epe = 6.717
This is 4-th epoch
total training loss = 6.572
total testing loss epe = 6.159
This is 5-th epoch
total training loss = 6.260
total testing loss epe = 6.436
This is 6-th epoch
total training loss = 6.053
total testing loss epe = 5.855
This is 7-th epoch
total training loss = 5.859
total testing loss epe = 6.019
This is 8-th epoch
total training loss = 5.705
total testing loss epe = 5.523
This is 9-th epoch
total training loss = 5.544
total testing loss epe = 5.337
This is 10-th epoch
total training loss = 5.435
total testing loss epe = 5.277
This is 11-th epoch
total training loss = 5.314
total testing loss epe = 5.542
This is 12-th epoch
 total training loss = 5.220
total testing loss epe = 5.126
This is 13-th epoch
total training loss = 5.110
total testing loss epe = 5.089
This is 14-th epoch
total training loss = 5.021
total testing loss epe = 5.431
This is 15-th epoch
total training loss = 4.955
total testing loss epe = 5.132
This is 16-th epoch
total training loss = 4.869
total testing loss epe = 5.052
This is 17-th epoch
total training loss = 4.791
total testing loss epe = 4.978
This is 18-th epoch
total training loss = 4.730
total testing loss epe = 4.915
This is 19-th epoch
total training loss = 4.669
total testing loss epe = 5.093
This is 20-th epoch
total training loss = 4.612
total testing loss epe = 4.875
This is 21-th epoch
total training loss = 4.544
total testing loss epe = 4.964

本来想发个sceneflow测试的图片，结果不知道怎么发图片，结果很不好，物体都不怎么完整，拿出几张图片，你将测试的结果输出来了吗？？

**Comment by changshuangzhao** [comment]
测试的EPE你的是多少？我也构建了其他的一些网络，最好的EPE=3.8左右，4个batch，训练了60轮，这种情况34000 / 4 * 60 = 510000次迭代，而论文中的训练十万次迭代以内就好了（iResNet），训练收敛的真快。

**Comment by changshuangzhao** [comment]
之前我做的网络也是，kitti 数据集>3px错误率7点多，但是图像显示出来的都是马赛克之类的，一点看不出来细节，只有大致的形状，不知道为什么

**Comment by changshuangzhao** [comment]
你们作业要求好高啊！！！！

**Comment by zhixuanli** [comment]
作业要求是真的高……也可能我太认真了，不想抄袭，所以就从头写了，结果发现坑太多了。

你可以用试试PSMNet预训练的模型，在两个数据集上测试一下，然后做一下可视化看看效果

另外，我一直不知道视察图如何变成深度图的，网络的输出结果就是视察图吗？

**Comment by changshuangzhao** [comment]
之前，PSPNet 预测网络输出是可以，不知道是不是后来版本让我改了，还是什么原因，用他训练好的数据，测试也是马赛克了，  输出来的结果就是视差图，深度图你要，z=fb/d,  f是焦距，b是两个相机之间的距离，d是视差，  fb数据库里会给你的

**Comment by zhixuanli** [comment]
我们这次的作业还要求提出改进方法，我琢磨的有这几方面，但是我只读了stereonet和GC net，可能我的这些想法早就被验证了：
1. cost volume计算方式由相减改为拼接（拼接通道），这样能尽可能多的保留信息。
2. attention module可以引入一下，因为不同位置的视察d并不同，注意力机制可能可以帮助网络关注于相同较大或较小的视察，帮助网络增强学习能力
3. 数据增强的重要性。之前我不太重视这方面，后来有篇语义分割领域的文章里做了对照试验，发现仅仅是把数据增强做的好一些，就能提升一个多的百分点，甚至更多。微软之前有篇文章就是把image复制了一下，然后拼接在一起，也就是2d变3d，然后就可以用3D卷积，然而就这种智障方法竟然能提升印象里大概有10个百分点……所以数据越多越好
4. 上采样的区别。双线性上采样（stereonet)所采用的，能够避免反卷积所带来的棋盘效应，但是双线性上采样并没有较好的学习能力，也没有引入下采样前的信息进行约束（像U-Net那样）
我的想法是用[亚像素卷积](http://www.cv-foundation.org/openaccess/content_cvpr_2016/papers/Shi_Real-Time_Single_Image_CVPR_2016_paper.pdf)作为上采样的方法，因为它是深度学习的方法，学习能力强，同时还能有效避免棋盘效应。然后可以再稍微改造一下，把ESPCN的网络中引入下采样前的原始分辨率的图像作为辅助（参考U-Net和ResNet的残差思想），应该会有一些提升。

以上内容纯属瞎想，做不出来别骂我😂😂😂

**Comment by changshuangzhao** [comment]
(1)cost volume  如何通过学习的方式获取D，在PSNet，GCNet都是固定D=192，这两个都是拼接，而且每一次移动都是所有通道进行移动，对于视差图，每个通道每一行的视差移动都是不一样的，如何去学习获得每一行的D，也就是说D不是固定的，且每一行的D也不同。想这么做，但一直没想到好方法。
（2）attention module。 这个可以提升网络，以前试过，获取初始视差图，然后用attention module细化
（3）数据增强。我看过的网络，3D网络都没有数据增强，而2D网络会有数据增强。但有的地方没想明白，对于颜色调整，亮度，色调，对比度，饱和度，只要在原始左右图像进行调整进行，不需要再视差图上，但对于平移，旋转，翻转我就不明白了，先说翻转，垂直翻转可以，左右图像视差图一起就行，水平翻转的话，对于同一个点，左右图视差图翻转后对应的视差就不对了，旋转也是，左右视差图旋转后，视差是对不上的
（4）这个没有想过，亚像素卷积不了解，还需看看

**Comment by zhixuanli** [comment]
cost volume计算完之后，每个像素点相当于有了多个“备选答案”，
注意看这行：[https://github.com/zhixuanli/StereoNet/blob/master/utils/cost_volume.py#L19](https://github.com/zhixuanli/StereoNet/blob/master/utils/cost_volume.py#L19)，for循环就是为了把每个可能的d值都尝试一下，把所有可能性都存起来（放在不同的通道里，不同通道对应不同的d）

soft arg min就是为了从多个答案中选出最好的那个

我理解的是，每个像素点都有不同的d已经实现了。

**Comment by qiuliang1993** [comment]
我也是刚了解这个关于深度的立体匹配，通过提取的左图和右图的特征后来求cost volume这一块的内容理解的不是很透彻，大神能讲解一些么？

**Comment by qiuliang1993** [comment]
我刚开始了解这个关于深度的立体匹配，我有一个问题是，在传统的方法中求cost volume时，是按照w方向来求视差的，但是在这篇论文中，我看了你的代码，让我不解的是，你是在H方向求的cost volume，我不太明白这一块，兄弟你能将将吗

**Comment by zhixuanli** [comment]
> 我刚开始了解这个关于深度的立体匹配，我有一个问题是，在传统的方法中求cost volume时，是按照w方向来求视差的，但是在这篇论文中，我看了你的代码，让我不解的是，你是在H方向求的cost volume，我不太明白这一块，兄弟你能将将吗

我去检查一下，可能我写错了😂😂

**Comment by zhixuanli** [comment]
@qiuliang1993 重新训练后，在kitti val上的小于3 pixel准确率是93.293，比以前高了3个百分点

**Comment by zhixuanli** [comment]
一直不知道怎么做可视化😂😂😂

回头学习一下

**Comment by ffeerrnn** [comment]
我想问下关于softmax部分，从公式上看，这个应该是softmin?
![image](https://user-images.githubusercontent.com/47510646/53002147-f1d3aa00-3466-11e9-9c73-119d6088928f.png)

**Comment by ffeerrnn** [comment]
hello，我想问下关于soft_argmin部分，从公式上看，应该是用的softmin吧，而且d的范围是[1,D]
![image](https://user-images.githubusercontent.com/47510646/53003704-8a1f5e00-346a-11e9-99e8-3f48bcfc47b0.png)
其实一直不太明白，文献中为什么会提到（D+1）/2K，(K次方，D我看你的代码里初始化为192)，因为D在下采样阶段没有用到，只是提取特征，所以为什么不是直接给出D（比如初始化为[0,11]）
![image](https://user-images.githubusercontent.com/47510646/53003736-9dcac480-346a-11e9-8844-16691e9fbfa6.png)

**Comment by zhixuanli** [comment]
@ffeerrnn 欢迎新开一个issue，这个实在太长了😂

明天上午回实验室看完给你回复😄

**Comment by zhixuanli** [comment]
@ffeerrnn 记得在StereoNet文章中，作者说用了3至4个下采样模块。这里不直接写成最终的搜索范围（0-11），是为了
1. 方便程序猿们理解得到搜索范围的过程
2. 为了适应不同数量K的下采样层，得到不同的搜索范围，避免每次都手动修改D的取值范围，所以直接设置为$(D+1)/2^K$。

**Comment by ffeerrnn** [comment]
hello,先谢谢哦，不过我的疑问重点不是在这里，1、soft_argmin给出的公式中，先对C(d)进行取负值，同样地在GCNet中也是这样，我想请问的是代码中是不是没有体现这一点，或是有其他什么解释；2、对于这个求和公式中，显然d的取值范围是[1,D]，而如果按照你的代码中的则变成了[0, (D+1)/2^K]
我是不太明白哈，包括论文里为什么是这样，所以劳烦解释下

**Comment by zhixuanli** [comment]
@ffeerrnn 谢谢你的指正，按照:[https://github.com/zyf12389/GC-Net/blob/master/gc_net.py](https://github.com/zyf12389/GC-Net/blob/master/gc_net.py)，的确应该是先取负值，而且应该在上采样后进行softarg min（d就会有更大的可选范围）。

当时参考的代码是[https://github.com/liuruijin17/GCNet-tensorflow/blob/master/gc_net.py#L66](https://github.com/liuruijin17/GCNet-tensorflow/blob/master/gc_net.py#L66)，这份代码中的softarg min没有取负值而且写的不够精简。

这就修改。

再次感谢！！

有时间我会把代码再认真检查一遍，的确有太多疏漏了。

**Comment by huangcimu** [comment]
> 之前我做的网络也是，kitti 数据集>3px错误率7点多，但是图像显示出来的都是马赛克之类的，一点看不出来细节，只有大致的形状，不知道为什么

你好，请问你是怎么可视化网络最后的视差图的

**Comment by huangcimu** [comment]
@changshuangzhao ，您好！请问如何可视化网络最后输出的视差图
